aptos-move/aptos-vm/src/aptos_vm.rs

```diff
@@ -2823,21 +3098,44 @@ pub(crate) fn is_account_init_for_sponsored_transaction(
     txn_data: &TransactionMetadata,
     features: &Features,
     resolver: &impl AptosMoveResolver,
+    module_storage: &impl AptosModuleStorage,
 ) -> VMResult<bool> {
-    Ok(
-        features.is_enabled(FeatureFlag::SPONSORED_AUTOMATIC_ACCOUNT_V1_CREATION)
-            && txn_data.fee_payer.is_some()
-            && txn_data.sequence_number == 0
-            && resolver
-                .get_resource_bytes_with_metadata_and_layout(
-                    &txn_data.sender(),
-                    &AccountResource::struct_tag(),
-                    &resolver.get_module_metadata(&AccountResource::struct_tag().module_id()),
-                    None,
-                )
-                .map(|(data, _)| data.is_none())
-                .map_err(|e| e.finish(Location::Undefined))?,
-    )
+    if features.is_enabled(FeatureFlag::SPONSORED_AUTOMATIC_ACCOUNT_V1_CREATION)
+        && txn_data.fee_payer.is_some()
+        && txn_data.sequence_number == 0
+    {
+        let metadata = fetch_module_metadata_for_struct_tag(
+            &AccountResource::struct_tag(),
+            features,
+            resolver,
+            module_storage,
+        )?;
+        let (maybe_bytes, _) = resolver
+            .get_resource_bytes_with_metadata_and_layout(
+                &txn_data.sender(),
+                &AccountResource::struct_tag(),
+                &metadata,
+                None,
+            )
+            .map_err(|e| e.finish(Location::Undefined))?;
+        return Ok(maybe_bytes.is_some());
+    }
+    Ok(false)
+}
+
+pub(crate) fn fetch_module_metadata_for_struct_tag(
```

Comment:
Maybe in some later PRs, a nice idea could be to separate a lot of these v1/v2 dispatchers into methods, move to a new file, and add some tests. Can be lower priority / clean-up after things are running e2e.